applications/trilinos_application/custom_utilities/trilinos_dof_updater.h

```diff
@@ -0,0 +1,240 @@
+    TrilinosDofUpdater& operator=(TrilinosDofUpdater const& rOther) = delete;
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    UniquePointer Create() const override
+    {
+        return UniquePointer(new TrilinosDofUpdater());
+    }
+
+    void Initialize(
+        DofsArrayType& rDofSet,
+        SystemVectorType& rDx) override
+    {
+
+        int system_size = TSparseSpace::Size(rDx);
+        int number_of_dofs = rDofSet.size();
+        std::vector< int > index_array(number_of_dofs);
+
+        // filling the array with the global ids
+        int counter = 0;
+        for(typename DofsArrayType::iterator i_dof = rDofSet.begin() ; i_dof != rDofSet.end() ; ++i_dof)
+        {
+            int id = i_dof->EquationId();
+            if( id < system_size )
+            {
+                index_array[counter++] = id;
+            }
+        }
+
+        std::sort(index_array.begin(),index_array.end());
+        std::vector<int>::iterator new_end = std::unique(index_array.begin(),index_array.end());
+        index_array.resize(new_end - index_array.begin());
+
+        int check_size = -1;
+        int tot_update_dofs = index_array.size();
+        rDx.Comm().SumAll(&tot_update_dofs,&check_size,1);
+        if ( (check_size < system_size) && (rDx.Comm().MyPID() == 0) )
+        {
+            std::stringstream msg;
+            msg << "Dof count is not correct. There are less dofs then expected." << std::endl;
+            msg << "Expected number of active dofs: " << system_size << ", dofs found: " << check_size << std::endl;
+            KRATOS_ERROR << msg.str();
+        }
+
+        // defining a map as needed
+        Epetra_Map dof_update_map(-1,index_array.size(), &(*(index_array.begin())),0,rDx.Comm() );
+
+        // defining the import instance
+        std::unique_ptr<Epetra_Import> p_dof_import(new Epetra_Import(dof_update_map,rDx.Map()));
+        mpDofImport.swap(p_dof_import);
+
+        mImportIsInitialized = true;
+    }
+
+    void Clear() override
+    {
+        mpDofImport.reset();
+        mImportIsInitialized = false;
+    }
+
+    void UpdateDOF(
+        DofsArrayType& rDofSet,
+        SystemVectorType& rDx) override
+    {
+        KRATOS_TRY;
+
+        if (!mImportIsInitialized)
+            this->Initialize(rDofSet,rDx);
+
+        int system_size = TSparseSpace::Size(rDx);
+
+        // defining a temporary vector to gather all of the values needed
+        Epetra_Vector local_dx( mpDofImport->TargetMap() );
+
+        // importing in the new temp vector the values
+        int ierr = local_dx.Import(rDx,*mpDofImport,Insert) ;
+        KRATOS_ERROR_IF(ierr != 0) << "Epetra failure found while trying to import Dx." << std::endl;
+
+        rDx.Comm().Barrier();
+
+        // performing the update
+        for (typename DofsArrayType::iterator it_dof = rDofSet.begin(); it_dof != rDofSet.end(); ++it_dof)
+        {
+            if (it_dof->IsFree()) {
+                int global_id = it_dof->EquationId();
```

Comment:
Be careful, with the new keys mixing integers and unsigned integers can originate serious troubles

Reply:
Hmm the ints were just copied from the original implementation in my bossak. I vaguely rememeber they were there because trilinos used them, but you are right that it is best if we can use size types instead. I will investigate.